Improved shared context. Refactor & simplify some specs.

marcandre opened the pull request:
I feel there are ways to clean some specs up with default rspec `let` groups.

How does this look for a start?

Stumbling upon this when working on #7868 because some specs are really abusing the API for autocorrection.

marcandre commented:
Or maybe I could put these with the `:config` shared environment instead?

Drenmi commented:
I think this looks like a great improvement to the spec seen in this PR. 👍

marcandre commented:
I have changed my PR. I believe this shared context is very flexible and would be a good replacement for `:config`.

bbatsov commented on spec/rubocop/formatter/clang_style_formatter_spec.rb:
```diff
@@ -1,6 +1,6 @@
 # frozen_string_literal: true
 
-RSpec.describe RuboCop::Formatter::ClangStyleFormatter do
+RSpec.describe RuboCop::Formatter::ClangStyleFormatter, :cop do
```
Maybe we can have a separate context for formatters?

marcandre replied:
We could, I haven't checked if there was extra things that could be factorized.

marcandre replied:
Didn't see much of interest

bbatsov commented:
I like the idea overall, but I'm not sure all the cops need so many things - you've picked to apply the suggest approach to two specs that are exactly typical cop specs. Perhaps we can get away with a smaller and more focused context for the actual cops.

bbatsov commented on lib/rubocop/rspec/shared_contexts.rb:
```diff
@@ -60,6 +60,67 @@
+
+  let(:source) { 'code = {some: :ruby}' }
+
+  let(:processed_source) { parse_source(source, 'test') }
```
Do cop tests often touch processed source and the source buffer?

marcandre replied:
Right, I will move `processed_source` in the second section since it is unlikely to be overriden. It is still useful as a shortcust to access the processed source though

marcandre commented:
Now that I'm more familiar with the `shared_context`, here's a revised PR.

It streamlines the `:config` shared context by:
1) Adding a `let(:cop)`
2) Removing the logic of `respond_to?(:cop_config)`. This could lead to cases where `let(:cop_config) {{}}` would have specs fail (see #8001).
3) Overrides `autocorrect` default setting (e.g. `JSONLoad`), not just the 'enabled' setting.

I simplified a few other specs. I also removed ~150 identical and redundant `let(:cop)`, so the PR might be a bit noisy. The main commit is the first one.

I'll open an different issue for things I dislike about our current `cop_helper` and `expect_offense`.

bbatsov commented:
Looks good! Thanks!
